File: javascript/ax-engine/test/client.test.js
import test from "node:test";
import assert from "node:assert/strict";
import http from "node:http";

import { AxEngineClient, AxEngineHttpError } from "../index.js";

async function withServer(handler, run) {
  const server = http.createServer(handler);
  await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
  const address = server.address();
  const baseUrl = `http://127.0.0.1:${address.port}`;
  try {
    await run(baseUrl);
  } finally {
    await new Promise((resolve, reject) =>
      server.close((error) => (error ? reject(error) : resolve())),
    );
  }
}

test("runtime fetches preview runtime metadata", async () => {
  await withServer((req, res) => {
    assert.equal(req.method, "GET");
    assert.equal(req.url, "/v1/runtime");
    res.setHeader("content-type", "application/json");
    res.end(
      JSON.stringify({
        service: "ax-engine-server",
        model_id: "qwen3_dense",
        runtime: {
          selected_backend: "ax_native",
        },
      }),
    );
  }, async (baseUrl) => {
    const client = new AxEngineClient({ baseUrl });
    const runtime = await client.runtime();
    assert.equal(runtime.model_id, "qwen3_dense");
    assert.equal(runtime.runtime.selected_backend, "ax_native");
  });
});

test("generate posts JSON to preview endpoint", async () => {
  await withServer((req, res) => {
    assert.equal(req.method, "POST");
    assert.equal(req.url, "/v1/generate");
    let body = "";
    req.setEncoding("utf8");
    req.on("data", (chunk) => {
      body += chunk;
    });
    req.on("end", () => {
      const payload = JSON.parse(body);
      assert.deepEqual(payload.input_tokens, [1, 2, 3]);
      assert.equal(payload.max_output_tokens, 2);
      res.setHeader("content-type", "application/json");
      res.end(
        JSON.stringify({
          request_id: 1,
          output_tokens: [4, 5],
        }),
      );
    });
  }, async (baseUrl) => {
    const client = new AxEngineClient({ baseUrl });
    const response = await client.generate({
      input_tokens: [1, 2, 3],
      max_output_tokens: 2,
    });
    assert.deepEqual(response.output_tokens, [4, 5]);
  });
});

test("streamGenerate parses named SSE events", async () => {
  await withServer((req, res) => {
    assert.equal(req.method, "POST");
    assert.equal(req.url, "/v1/generate/stream");
    res.writeHead(200, {
      "content-type": "text/event-stream",
      "cache-control": "no-cache",
      connection: "keep-alive",
    });
    res.write('event: request\ndata: {"request_id":1}\n\n');
    res.write(
      'event: step\ndata: {"delta_tokens":[42],"delta_token_logprobs":[-0.25],"step":{"metal_dispatch":{"runtime_model_conditioned_inputs":true,"runtime_real_model_tensor_inputs":true,"runtime_complete_model_forward_supported":true,"runtime_model_family":"qwen3_dense","execution_direct_decode_token_count":1,"execution_direct_decode_checksum_lo":4660,"execution_logits_output_count":1,"execution_remaining_logits_handle_count":0,"execution_model_bound_ffn_decode":true,"execution_real_model_forward_completed":true,"execution_prefix_native_dispatch_count":35,"execution_prefix_cpu_reference_dispatch_count":1,"execution_qkv_projection_token_count":72,"execution_layer_continuation_token_count":37,"execution_logits_projection_token_count":1,"execution_logits_vocab_scan_row_count":151936}}}\n\n',
    );
    res.end('event: response\ndata: {"finish_reason":"max_tokens"}\n\n');
  }, async (baseUrl) => {
    const client = new AxEngineClient({ baseUrl });
    const events = [];
    for await (const event of client.streamGenerate({
      input_tokens: [1, 2, 3],
      max_output_tokens: 1,
    })) {
      events.push(event);
    }

    assert.deepEqual(events, [
      { event: "request", data: { request_id: 1 } },
      {
        event: "step",
        data: {
          delta_tokens: [42],
          delta_token_logprobs: [-0.25],
          step: {
            metal_dispatch: {
              runtime_model_conditioned_inputs: true,
              runtime_real_model_tensor_inputs: true,
              runtime_complete_model_forward_supported: true,
              runtime_model_family: "qwen3_dense",
              execution_direct_decode_token_count: 1,
              execution_direct_decode_checksum_lo: 4660,
              execution_logits_output_count: 1,
              execution_remaining_logits_handle_count: 0,
              execution_model_bound_ffn_decode: true,
              execution_real_model_forward_completed: true,
              execution_prefix_native_dispatch_count: 35,
              execution_prefix_cpu_reference_dispatch_count: 1,
              execution_qkv_projection_token_count: 72,
              execution_layer_continuation_token_count: 37,
              execution_logits_projection_token_count: 1,
              execution_logits_vocab_scan_row_count: 151936,
            },
          },
        },
      },
      { event: "response", data: { finish_reason: "max_tokens" } },
    ]);
  });
});

test("streamCompletion stops on OpenAI [DONE] sentinel", async () => {
  await withServer((req, res) => {
    assert.equal(req.method, "POST");
    assert.equal(req.url, "/v1/completions");
    res.writeHead(200, {
      "content-type": "text/event-stream",
      "cache-control": "no-cache",
      connection: "keep-alive",
    });
    res.write('data: {"id":"cmpl-1","choices":[{"text":"Hello"}]}\n\n');
    res.end("data: [DONE]\n\n");
  }, async (baseUrl) => {
    const client = new AxEngineClient({ baseUrl });
    const events = [];
    for await (const event of client.streamCompletion({
      prompt: "Hello",
      model: "qwen3_dense",
    })) {
      events.push(event);
    }

    assert.deepEqual(events, [
      {
        event: "message",
        data: {
          id: "cmpl-1",
          choices: [{ text: "Hello" }],
        },
      },
    ]);
  });
});

test("embeddings posts token array to OpenAI-shaped endpoint", async () => {
  await withServer((req, res) => {
    assert.equal(req.method, "POST");
    assert.equal(req.url, "/v1/embeddings");
    let body = "";
    req.setEncoding("utf8");
    req.on("data", (chunk) => {
      body += chunk;
    });
    req.on("end", () => {
      const payload = JSON.parse(body);
      assert.deepEqual(payload.input, [1, 2, 3]);
      assert.equal(payload.pooling, "last");
      assert.equal(payload.normalize, true);
      res.setHeader("content-type", "application/json");
      res.end(
        JSON.stringify({
          object: "list",
          data: [{ object: "embedding", embedding: [0.1, 0.2], index: 0 }],
          model: "qwen3_embedding",
          usage: { prompt_tokens: 3, total_tokens: 3 },
        }),
      );
    });
  }, async (baseUrl) => {
    const client = new AxEngineClient({ baseUrl });
    const response = await client.embeddings({
      model: "qwen3_embedding",
      input: [1, 2, 3],
      pooling: "last",
      normalize: true,
    });
    assert.deepEqual(response.data[0].embedding, [0.1, 0.2]);
    assert.equal(response.usage.total_tokens, 3);
  });
});

test("completion returns OpenAI-shaped response", async () => {
  await withServer((req, res) => {
    assert.equal(req.method, "POST");
    assert.equal(req.url, "/v1/completions");
    res.setHeader("content-type", "application/json");
    res.end(
      JSON.stringify({
        id: "cmpl-1",
        object: "text_completion",
        created: 1234567890,
        model: "qwen3_dense",
        choices: [{ index: 0, text: "Hello world", finish_reason: "stop" }],
        usage: { prompt_tokens: 3, completion_tokens: 2, total_tokens: 5 },
      }),
    );
  }, async (baseUrl) => {
    const client = new AxEngineClient({ baseUrl });
    const response = await client.completion({ prompt: "Hello", max_tokens: 32 });
    assert.equal(response.object, "text_completion");
    assert.equal(response.choices[0].text, "Hello world");
    assert.equal(response.usage.total_tokens, 5);
  });
});

test("chatCompletion returns OpenAI-shaped response", async () => {
  await withServer((req, res) => {
    assert.equal(req.method, "POST");
    assert.equal(req.url, "/v1/chat/completions");
    let body = "";
    req.setEncoding("utf8");
    req.on("data", (chunk) => { body += chunk; });
    req.on("end", () => {
      const payload = JSON.parse(body);
      assert.deepEqual(payload.messages[0], { role: "user", content: "Hello!" });
      res.setHeader("content-type", "application/json");
      res.end(
        JSON.stringify({
          id: "chatcmpl-1",
          object: "chat.completion",
          created: 1234567890,
          model: "qwen3_dense",
          choices: [
            {
              index: 0,
              message: { role: "assistant", content: "Hi there!" },
              finish_reason: "stop",
            },
          ],
          usage: { prompt_tokens: 5, completion_tokens: 3, total_tokens: 8 },
        }),
      );
    });
  }, async (baseUrl) => {
    const client = new AxEngineClient({ baseUrl });
    const response = await client.chatCompletion({
      messages: [{ role: "user", content: "Hello!" }],
      max_tokens: 32,
    });
    assert.equal(response.object, "chat.completion");
    assert.equal(response.choices[0].message.content, "Hi there!");
    assert.equal(response.usage.completion_tokens, 3);
  });
});

test("non-2xx responses raise AxEngineHttpError with payload", async () => {
  await withServer((_req, res) => {
    res.writeHead(400, {
      "content-type": "application/json",
    });
    res.end(
      JSON.stringify({
        error: {
          code: "invalid_request",
          message: "bad request",
        },
      }),
    );
  }, async (baseUrl) => {
    const client = new AxEngineClient({ baseUrl });
    await assert.rejects(
      () => client.generate({ input_tokens: [1] }),
      (error) => {
        assert.ok(error instanceof AxEngineHttpError);
        assert.equal(error.status, 400);
        assert.equal(error.message, "bad request");
        assert.deepEqual(error.payload, {
          error: {
            code: "invalid_request",
            message: "bad request",
          },
        });
        return true;
      },
    );
  });
});

File: sdk/go/axengine/client_test.go
package axengine

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"testing"
)

// startServer launches a temporary HTTP server, calls run with its base URL,
// then shuts it down.
func startServer(t *testing.T, mux *http.ServeMux, run func(baseURL string)) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	srv := &http.Server{Handler: mux}
	go srv.Serve(ln) //nolint:errcheck
	t.Cleanup(func() { srv.Close() })
	run(fmt.Sprintf("http://127.0.0.1:%d", ln.Addr().(*net.TCPAddr).Port))
}

func writeJSON(w http.ResponseWriter, v any) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(v) //nolint:errcheck
}

func TestHealthOK(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			t.Errorf("method: got %s want GET", r.Method)
		}
		writeJSON(w, HealthResponse{Status: "ok", Service: "ax-engine-server", ModelID: "qwen3_dense"})
	})
	startServer(t, mux, func(baseURL string) {
		client := NewClient(&ClientOptions{BaseURL: baseURL})
		resp, err := client.Health(context.Background())
		if err != nil {
			t.Fatal(err)
		}
		if resp.Status != "ok" {
			t.Errorf("status: got %q want %q", resp.Status, "ok")
		}
		if resp.ModelID != "qwen3_dense" {
			t.Errorf("model_id: got %q want %q", resp.ModelID, "qwen3_dense")
		}
	})
}

func TestCompletionOK(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/completions", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			t.Errorf("method: got %s want POST", r.Method)
		}
		var req OpenAiCompletionRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Errorf("decode body: %v", err)
		}
		writeJSON(w, OpenAiCompletionResponse{
			ID:     "cmpl-1",
			Object: "text_completion",
			Model:  "qwen3_dense",
			Choices: []OpenAiCompletionChoice{
				{Index: 0, Text: "Hello world", FinishReason: Ptr("stop")},
			},
			Usage: &OpenAiUsage{PromptTokens: 3, CompletionTokens: 2, TotalTokens: 5},
		})
	})
	startServer(t, mux, func(baseURL string) {
		client := NewClient(&ClientOptions{BaseURL: baseURL})
		resp, err := client.Completion(context.Background(), OpenAiCompletionRequest{
			Prompt:    "Hello",
			MaxTokens: Ptr(32),
		})
		if err != nil {
			t.Fatal(err)
		}
		if resp.Object != "text_completion" {
			t.Errorf("object: got %q", resp.Object)
		}
		if resp.Choices[0].Text != "Hello world" {
			t.Errorf("text: got %q", resp.Choices[0].Text)
		}
		if resp.Usage.TotalTokens != 5 {
			t.Errorf("total_tokens: got %d want 5", resp.Usage.TotalTokens)
		}
	})
}

func TestChatCompletionOK(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/chat/completions", func(w http.ResponseWriter, r *http.Request) {
		var req OpenAiChatCompletionRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Errorf("decode body: %v", err)
		}
		if len(req.Messages) != 1 {
			t.Errorf("messages: got %d want 1", len(req.Messages))
		}
		writeJSON(w, OpenAiChatCompletionResponse{
			ID:     "chatcmpl-1",
			Object: "chat.completion",
			Model:  "qwen3_dense",
			Choices: []OpenAiChatCompletionChoice{
				{
					Index:        0,
					Message:      OpenAiChatMessageResponse{Role: "assistant", Content: "Hi there!"},
					FinishReason: Ptr("stop"),
				},
			},
			Usage: &OpenAiUsage{PromptTokens: 5, CompletionTokens: 3, TotalTokens: 8},
		})
	})
	startServer(t, mux, func(baseURL string) {
		client := NewClient(&ClientOptions{BaseURL: baseURL})
		resp, err := client.ChatCompletion(context.Background(), OpenAiChatCompletionRequest{
			Messages:  []OpenAiChatMessage{{Role: "user", Content: "Hello!"}},
			MaxTokens: Ptr(64),
		})
		if err != nil {
			t.Fatal(err)
		}
		if resp.Choices[0].Message.Content != "Hi there!" {
			t.Errorf("content: got %q", resp.Choices[0].Message.Content)
		}
		if resp.Usage.CompletionTokens != 3 {
			t.Errorf("completion_tokens: got %d want 3", resp.Usage.CompletionTokens)
		}
	})
}

func TestHTTPErrorPropagated(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/completions", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprint(w, `{"error":{"message":"bad request"}}`)
	})
	startServer(t, mux, func(baseURL string) {
		client := NewClient(&ClientOptions{BaseURL: baseURL})
		_, err := client.Completion(context.Background(), OpenAiCompletionRequest{Prompt: "x"})
		if err == nil {
			t.Fatal("expected error")
		}
		httpErr, ok := err.(*HTTPError)
		if !ok {
			t.Fatalf("expected *HTTPError, got %T", err)
		}
		if httpErr.Status != http.StatusBadRequest {
			t.Errorf("status: got %d want 400", httpErr.Status)
		}
		if httpErr.Message != "bad request" {
			t.Errorf("message: got %q want %q", httpErr.Message, "bad request")
		}
	})
}

func TestStreamChatCompletionOK(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/chat/completions", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")
		flusher, ok := w.(http.Flusher)
		if !ok {
			t.Error("ResponseWriter does not implement http.Flusher")
			return
		}
		chunks := []string{
			`data: {"id":"c1","choices":[{"index":0,"delta":{"content":"Hello"},"finish_reason":null}]}`,
			`data: {"id":"c1","choices":[{"index":0,"delta":{"content":" world"},"finish_reason":"stop"}]}`,
			"data: [DONE]",
		}
		for _, chunk := range chunks {
			fmt.Fprintf(w, "%s\n\n", chunk)
			flusher.Flush()
		}
	})
	startServer(t, mux, func(baseURL string) {
		client := NewClient(&ClientOptions{BaseURL: baseURL})
		ch, errCh := client.StreamChatCompletion(context.Background(), OpenAiChatCompletionRequest{
			Messages:  []OpenAiChatMessage{{Role: "user", Content: "Hi"}},
			MaxTokens: Ptr(16),
		})

		var texts []string
		for chunk := range ch {
			if len(chunk.Choices) > 0 && chunk.Choices[0].Delta.Content != nil {
				texts = append(texts, *chunk.Choices[0].Delta.Content)
			}
		}
		if err := <-errCh; err != nil {
			t.Fatal(err)
		}

		if len(texts) != 2 {
			t.Fatalf("chunks: got %d want 2", len(texts))
		}
		if texts[0] != "Hello" || texts[1] != " world" {
			t.Errorf("text: got %v", texts)
		}
	})
}

func TestStreamCompletionOK(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/completions", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/event-stream")
		flusher := w.(http.Flusher)
		fmt.Fprint(w, "data: {\"id\":\"cmpl-1\",\"choices\":[{\"index\":0,\"text\":\"Once\",\"finish_reason\":null}]}\n\n")
		flusher.Flush()
		fmt.Fprint(w, "data: [DONE]\n\n")
		flusher.Flush()
	})
	startServer(t, mux, func(baseURL string) {
		client := NewClient(&ClientOptions{BaseURL: baseURL})
		ch, errCh := client.StreamCompletion(context.Background(), OpenAiCompletionRequest{
			Prompt:    "Once upon",
			MaxTokens: Ptr(8),
		})
		var combined string
		for chunk := range ch {
			combined += chunk.Choices[0].Text
		}
		if err := <-errCh; err != nil {
			t.Fatal(err)
		}
		if combined != "Once" {
			t.Errorf("text: got %q want %q", combined, "Once")
		}
	})
}

func TestStreamGenerateOK(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/generate/stream", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			t.Errorf("method: got %s want POST", r.Method)
		}
		if r.Header.Get("Accept") != "text/event-stream" {
			t.Errorf("accept: got %q want text/event-stream", r.Header.Get("Accept"))
		}
		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")
		flusher := w.(http.Flusher)
		fmt.Fprint(w, "event: request\ndata: {\"request\":{\"request_id\":1,\"model_id\":\"qwen3_dense\",\"state\":\"active\",\"prompt_tokens\":[1,2,3],\"processed_prompt_tokens\":3,\"output_tokens\":[],\"prompt_len\":3,\"output_len\":0,\"max_output_tokens\":4,\"cancel_requested\":false,\"route\":{}}}\n\n")
		flusher.Flush()
		fmt.Fprint(w, "event: step\ndata: {\"request\":{\"request_id\":1,\"model_id\":\"qwen3_dense\",\"state\":\"active\",\"prompt_tokens\":[1,2,3],\"processed_prompt_tokens\":3,\"output_tokens\":[42],\"prompt_len\":3,\"output_len\":1,\"max_output_tokens\":4,\"cancel_requested\":false,\"route\":{}},\"step\":{\"scheduled_requests\":1,\"scheduled_tokens\":1,\"ttft_events\":1,\"prefix_hits\":0,\"kv_usage_blocks\":1,\"evictions\":0,\"cpu_time_us\":100,\"runner_time_us\":200},\"delta_tokens\":[42],\"delta_text\":\"hello\"}\n\n")
		flusher.Flush()
		fmt.Fprint(w, "event: response\ndata: {\"response\":{\"request_id\":1,\"model_id\":\"qwen3_dense\",\"prompt_tokens\":[1,2,3],\"output_tokens\":[42],\"status\":\"finished\",\"finish_reason\":\"stop\",\"step_count\":1,\"route\":{}}}\n\n")
		flusher.Flush()
	})
	startServer(t, mux, func(baseURL string) {
		client := NewClient(&ClientOptions{BaseURL: baseURL})
		ch, errCh := client.StreamGenerate(context.Background(), PreviewGenerateRequest{
			InputTokens:     []int{1, 2, 3},
			MaxOutputTokens: Ptr(4),
		})

		var events []GenerateStreamEvent
		for ev := range ch {
			events = append(events, ev)
		}
		if err := <-errCh; err != nil {
			t.Fatal(err)
		}

		if len(events) != 3 {
			t.Fatalf("events: got %d want 3", len(events))
		}
		if events[0].Event != "request" || events[0].Request == nil {
			t.Errorf("event[0]: got event=%q request=%v", events[0].Event, events[0].Request)
		}
		if events[1].Event != "step" || events[1].Step == nil {
			t.Errorf("event[1]: got event=%q step=%v", events[1].Event, events[1].Step)
		}
		if events[1].Step != nil {
			step := events[1].Step
			if len(step.DeltaTokens) != 1 || step.DeltaTokens[0] != 42 {
				t.Errorf("step delta_tokens: got %v", step.DeltaTokens)
			}
			if step.DeltaText == nil || *step.DeltaText != "hello" {
				t.Errorf("step delta_text: got %v", step.DeltaText)
			}
		}
		if events[2].Event != "response" || events[2].Response == nil {
			t.Errorf("event[2]: got event=%q response=%v", events[2].Event, events[2].Response)
		}
		if events[2].Response != nil && events[2].Response.Response.FinishReason != nil {
			if *events[2].Response.Response.FinishReason != "stop" {
				t.Errorf("finish_reason: got %q want stop", *events[2].Response.Response.FinishReason)
			}
		}
	})
}

func TestContextCancellation(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/chat/completions", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/event-stream")
		// never writes — blocks until client disconnects
		<-r.Context().Done()
	})
	startServer(t, mux, func(baseURL string) {
		ctx, cancel := context.WithCancel(context.Background())
		client := NewClient(&ClientOptions{BaseURL: baseURL})
		ch, errCh := client.StreamChatCompletion(ctx, OpenAiChatCompletionRequest{
			Messages: []OpenAiChatMessage{{Role: "user", Content: "x"}},
		})
		cancel()
		// drain channels — should not block
		for range ch {
		}
		<-errCh
	})
}
